File: dataforge-storage/src/main/java/hep/dataforge/storage/filestorage/FileObjectLoader.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package hep.dataforge.storage.filestorage;

import hep.dataforge.exceptions.StorageException;
import hep.dataforge.meta.Meta;
import hep.dataforge.storage.api.Storage;
import hep.dataforge.storage.loaders.AbstractBinaryLoader;
import org.apache.commons.io.FilenameUtils;

import java.io.ByteArrayOutputStream;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.Serializable;
import java.nio.file.Path;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

/**
 * A file loader to store Serializable java objects
 *
 * @author Alexander Nozik
 */
public class FileObjectLoader<T extends Serializable> extends AbstractBinaryLoader<T> {

    public static <T extends Serializable> FileObjectLoader<T> fromEnvelope(Storage storage, FileEnvelope envelope) throws Exception {
        if (FileStorageEnvelopeType.validate(envelope, OBJECT_LOADER_TYPE)) {
            FileObjectLoader res = new FileObjectLoader(storage,
                    FilenameUtils.getBaseName(envelope.getFile().getFileName().toString()),
                    envelope.meta(),
                    envelope.getFile());
            res.setReadOnly(envelope.isReadOnly());
            return res;
        } else {
            throw new StorageException("Is not a valid object loader file");
        }
    }

    private final Path path;
    private final Map<String, T> dataMap = new HashMap<>();

    /**
     * An file used for pushing
     */
    private FileEnvelope file;

    public FileObjectLoader(Storage storage, String name, Meta meta, Path path) {
        super(storage, name, meta);
        this.path = path;
    }

    @Override
    public void open() throws Exception {
        if (this.meta == null) {
            this.meta = getFile().meta();
        }
        if (!isOpen()) {
            file = buildEnvelope(isReadOnly());
        }
    }

    @Override
    public boolean isOpen() {
        return file != null;
    }

    @Override
    public void close() throws Exception {
        getFile().close();
        file = null;
        super.close();
    }

    private FileEnvelope getFile() throws Exception {
        if (file == null) {
            open();
        }
        return file;
    }

    @Override
    public boolean isEmpty() {
        try {
            return !getFile().hasData();
        } catch (Exception ex) {
            throw new RuntimeException("Can't access loader envelope", ex);
        }
    }

    public Map<String, T> getDataMap() throws StorageException {
        if (dataMap.isEmpty()) {
            dataMap.putAll(readDataMap());
        }
        return dataMap;
    }

    @SuppressWarnings("unchecked")
    protected synchronized Map<String, T> readDataMap() throws StorageException {
        try (ObjectInputStream ois = new ObjectInputStream(getEnvelope().getData().getStream())) {
            return (Map<String, T>) ois.readObject();
        } catch (Exception ex) {
            return new HashMap<>();
            //throw new StorageException(ex);
        }
    }

    protected synchronized void writeDataMap(Map<String, T> data) throws StorageException {
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        try (ObjectOutputStream oos = new ObjectOutputStream(baos)) {
            oos.writeObject(data);
            getEnvelope().clearData();
            getEnvelope().append(baos.toByteArray());
        } catch (Exception ex) {
            throw new StorageException(ex);
        }
    }

    private FileEnvelope getEnvelope() throws StorageException {
        if (this.file == null) {
            this.file = buildEnvelope(false);
        }
        return this.file;
    }

    private FileEnvelope buildEnvelope(boolean readOnly) {
        return new FileEnvelope(path, readOnly);
    }

    @Override
    public Collection<String> fragmentNames() {
        try {
            return getDataMap().keySet();
        } catch (Exception ex) {
            return Collections.emptyList();
        }
    }

    @Override
    public T pull(String fragmentName) throws StorageException {
        return getDataMap().get(fragmentName);
    }

    @Override
    public void push(String fragmentName, T data) throws StorageException {
        getDataMap().put(fragmentName, data);
        writeDataMap(dataMap);
    }

}

File: dataforge-storage/src/main/kotlin/hep/dataforge/storage/filestorage/FileObjectLoader.kt
/*
 * Copyright  2018 Alexander Nozik.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package hep.dataforge.storage.filestorage

import hep.dataforge.exceptions.StorageException
import hep.dataforge.meta.Meta
import hep.dataforge.storage.api.ObjectLoader
import hep.dataforge.storage.api.Storage
import org.apache.commons.io.FilenameUtils
import java.io.ByteArrayOutputStream
import java.io.ObjectInputStream
import java.io.ObjectOutputStream
import java.io.Serializable
import java.util.*

/**
 * A file loader to store Serializable java objects
 *
 * @author Alexander Nozik
 */
class FileObjectLoader<T : Serializable>(storage: Storage, name: String, meta: Meta, file:FileEnvelope) : FileLoader(storage, name, meta, file), ObjectLoader<T> {
    private val dataMap = HashMap<String, T>()

    @Throws(StorageException::class)
    fun getDataMap(): MutableMap<String, T> {
        if (dataMap.isEmpty()) {
            dataMap.putAll(readDataMap())
        }
        return dataMap
    }

    @Synchronized
    @Throws(StorageException::class)
    protected fun readDataMap(): Map<String, T> {
        try {
            ObjectInputStream(file.data.stream).use { ois -> return ois.readObject() as Map<String, T> }
        } catch (ex: Exception) {
            return HashMap()
            //throw new StorageException(ex);
        }

    }

    @Synchronized
    @Throws(StorageException::class)
    protected fun writeDataMap(data: Map<String, T>) {
        val baos = ByteArrayOutputStream()
        try {
            ObjectOutputStream(baos).use { oos ->
                oos.writeObject(data)
                file.clearData()
                file.append(baos.toByteArray())
            }
        } catch (ex: Exception) {
            throw StorageException(ex)
        }

    }
    override fun fragmentNames(): Collection<String> {
        try {
            return getDataMap().keys
        } catch (ex: Exception) {
            return emptyList()
        }

    }

    @Throws(StorageException::class)
    override fun pull(fragmentName: String): T {
        return getDataMap()[fragmentName]
    }

    @Throws(StorageException::class)
    override fun push(fragmentName: String, data: T) {
        getDataMap()[fragmentName] = data
        writeDataMap(dataMap)
    }

    companion object {

        @Throws(Exception::class)
        fun <T : Serializable> fromEnvelope(storage: Storage, envelope: FileEnvelope): FileObjectLoader<T> {
            if (FileStorageEnvelopeType.validate(envelope, ObjectLoader.OBJECT_LOADER_TYPE)) {
                val res = FileObjectLoader(storage,
                        FilenameUtils.getBaseName(envelope.file.fileName.toString()),
                        envelope.meta,
                        envelope.file)
                res.isReadOnly = envelope.isReadOnly
                return res
            } else {
                throw StorageException("Is not a valid object loader file")
            }
        }
    }

}
